plugin/testing_write_upsert.go

```diff
@@ -132,3 +132,74 @@ func (s *WriterTestSuite) testUpsertAll(ctx context.Context) error {
 
 	return nil
 }
+
+func (s *WriterTestSuite) testInsertDuplicatePK(ctx context.Context) error {
+	const rowsPerRecord = 10
+	tableName := s.tableNameForTest("upsert_duplicate_pk")
+	table := schema.TestTable(tableName, s.genDatOptions)
+	table.Columns.Get("id").PrimaryKey = true
+	if err := s.plugin.writeOne(ctx, &message.WriteMigrateTable{
+		Table: table,
+	}); err != nil {
+		return fmt.Errorf("failed to create table: %w", err)
+	}
+	tg := schema.NewTestDataGenerator(0)
+
+	normalRecord := tg.Generate(table, schema.GenTestDataOptions{
+		MaxRows:            rowsPerRecord,
+		TimePrecision:      s.genDatOptions.TimePrecision,
+		UseHomogeneousType: s.useHomogeneousTypes,
+		DuplicateID:        true,
+	})
+
+	// normalRecord
+	if err := s.plugin.writeOne(ctx, &message.WriteInsert{
+		Record: normalRecord,
+	}); err != nil {
+		return fmt.Errorf("failed to insert record: %w", err)
+	}
+
+	records, err := s.plugin.readAll(ctx, table)
+	if err != nil {
+		return fmt.Errorf("failed to readAll: %w", err)
+	}
+	sortRecords(table, records, "id")
+
+	totalItems := TotalRows(records)
+	if totalItems != 1 {
+		return fmt.Errorf("expected items after initial insert: %d, got %d", 1, totalItems)
+	}
+
+	if diff := RecordsDiff(table.ToArrowSchema(), records, []arrow.Record{extractLastRowFromRecord(table, normalRecord)}); diff != "" {
+		return fmt.Errorf("record differs after insert: %s", diff)
+	}
+
+	return nil
+}
+
+func extractLastRowFromRecord(table *schema.Table, existingRecord arrow.Record) arrow.Record {
+	sc := table.ToArrowSchema()
+	var lastRecord []arrow.Record
+	bldr := array.NewRecordBuilder(memory.DefaultAllocator, sc)
+	for i, c := range table.Columns {
+		col := existingRecord.Column(i)
+		err := bldr.Field(i).AppendValueFromString(col.ValueStr(int(existingRecord.NumRows()) - 1))
```

Comment:
Could we extract `col.ValueStr(int(existingRecord.NumRows()) - 1`? It's used a bit below too